[Painless Lab] NP migration

alisonelizabeth opened the pull request:
This PR moves the painless lab plugin over to the new platform.

### How to test
I'm not as familiar with this app as others might be, so please in general verify the app still behaves as it did prior to the migration. 

- Confirm painless scripts execute as expected
- Verify "Painless Lab" appears in the Kibana plugins directory on the home page
- Verify setting `xpack.painless_lab.enabled: false` disables the app
- Verify SCSS still working as expected

_Note:_ The license handling logic was brought over from [Seb's implementation in index_management](https://github.com/elastic/kibana/blob/master/x-pack/plugins/index_management/server/services/license.ts).

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

alisonelizabeth commented on x-pack/plugins/painless_lab/public/plugin.tsx:
```diff
@@ -0,0 +1,99 @@
+      ) as any,
+      enableRouting: false,
+      disabled: false,
+      // tooltipContent: xpackInfo.get('features.painlessLab.message'),
```
I'm not sure what this is for or how to test (it was brought over from the previous code). @jloleysens or @cjcenizal - any ideas?

alisonelizabeth replied:
Ah, I see. I'm going to delete this for now.

Currently, the app is redirecting back to /dev-tools if the license is invalid. This is [current functionality](https://github.com/elastic/kibana/blob/app/painless/x-pack/legacy/plugins/painless_lab/public/register.tsx#L64) that I brought over. It looks like this was originally copied over from `grokdebugger`, as I see it is doing the same thing.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/)
* Commit: b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/","number":"32164","commit":"b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/)
* Commit: d5cf2b358b8e06d02b32cd766f41079bf2f2c188

### History
* :broken_heart: [Build #32164](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/) failed b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/","number":"32164","commit":"b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/","number":"32200","commit":"d5cf2b358b8e06d02b32cd766f41079bf2f2c188"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/)
* Commit: c0c0b8e0a5d56fe2db59964686d2cdf907d3d008

### History
* :broken_heart: [Build #32200](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/) failed d5cf2b358b8e06d02b32cd766f41079bf2f2c188
* :broken_heart: [Build #32164](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/) failed b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/","number":"32164","commit":"b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/","number":"32200","commit":"d5cf2b358b8e06d02b32cd766f41079bf2f2c188"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/","number":"32269","commit":"c0c0b8e0a5d56fe2db59964686d2cdf907d3d008"}]}
PIPELINE-->

jloleysens commented on x-pack/plugins/painless_lab/server/routes/api/execute.ts:
```diff
@@ -0,0 +1,47 @@
+      },
+    },
+    license.guardApiRoute(async (ctx, req, res) => {
+      const body = req.body as typeof bodySchema.type;
```
I've also run into this 😄. The higher order function (in this case `license.guardApiRoute`) needs to accept the three RequestHandler generic params `<P, Q, B>` and pass them through to its own handler and request object.

Like here:
https://github.com/elastic/kibana/blob/babf81bdc02174145f0416e579f8e18a8d27c2ed/x-pack/plugins/watcher/server/lib/license_pre_routing_factory/license_pre_routing_factory.ts#L15

We can also review this down the line.

jloleysens commented on x-pack/plugins/painless_lab/common/constants.ts:
```diff
@@ -0,0 +1,19 @@
+export const PLUGIN = {
+  id: 'painlessLab',
+  minimumLicenseType: basicLicense,
+  getI18nName: (i18n: any): string =>
```
I can't see where this is being used?

alisonelizabeth commented:
@jloleysens I addressed your feedback and fixed the issue when a user submits an invalid painless script (great catch btw!). The UI is actually expecting a `200` in this scenario ([related code](https://github.com/elastic/kibana/blob/app/painless/x-pack/legacy/plugins/painless_lab/public/lib/helpers.ts#L74)). Mind taking another look?

jloleysens commented on x-pack/plugins/painless_lab/server/routes/api/execute.ts:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        if (isEsError(e)) {
+          return res.customError({ statusCode: e.statusCode, body: e });
```
I'm just wondering whether we can't collapse this section into this if check:

```ts
	// Invalid painless script was submitted
        // Return 200 with error object
        if (isEsError(e)) {
          return res.ok({ body: e.body });
        }
```

I'm guessing if we detect an ES error here it means that something went wrong with the call to painless script execution endpoint.

alisonelizabeth replied:
That's a valid point! Fixed.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32415/)
* Commit: c76542fe057b8a2b8701ca403a3f3815251b936c

### History
* :green_heart: [Build #32269](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/) succeeded c0c0b8e0a5d56fe2db59964686d2cdf907d3d008
* :broken_heart: [Build #32200](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/) failed d5cf2b358b8e06d02b32cd766f41079bf2f2c188
* :broken_heart: [Build #32164](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/) failed b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/","number":"32164","commit":"b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/","number":"32200","commit":"d5cf2b358b8e06d02b32cd766f41079bf2f2c188"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/","number":"32269","commit":"c0c0b8e0a5d56fe2db59964686d2cdf907d3d008"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32415/","number":"32415","commit":"c76542fe057b8a2b8701ca403a3f3815251b936c"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32569/)
* Commit: 54b5d15aacd2c10f520a846acaf9eef0fd187fb1

### History
* :green_heart: [Build #32415](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32415/) succeeded c76542fe057b8a2b8701ca403a3f3815251b936c
* :green_heart: [Build #32269](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/) succeeded c0c0b8e0a5d56fe2db59964686d2cdf907d3d008
* :broken_heart: [Build #32200](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/) failed d5cf2b358b8e06d02b32cd766f41079bf2f2c188
* :broken_heart: [Build #32164](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/) failed b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/","number":"32164","commit":"b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/","number":"32200","commit":"d5cf2b358b8e06d02b32cd766f41079bf2f2c188"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/","number":"32269","commit":"c0c0b8e0a5d56fe2db59964686d2cdf907d3d008"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32415/","number":"32415","commit":"c76542fe057b8a2b8701ca403a3f3815251b936c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32569/","number":"32569","commit":"54b5d15aacd2c10f520a846acaf9eef0fd187fb1"}]}
PIPELINE-->

cjcenizal commented on x-pack/plugins/painless_lab/server/routes/api/execute.ts:
```diff
@@ -0,0 +1,51 @@
+
+const bodySchema = schema.object({
+  script: schema.object({
+    source: schema.string(),
```
We'll need to update this schema to support the other parameters that will be provided if the user specifies "Parameters" or the other options provided by the other "Context" choices. For example, if you specify "Parameters" as in the screenshot below the request fails because it's invalidated by this schema:

![image](https://user-images.githubusercontent.com/1238659/76475229-6f908d80-63bb-11ea-9471-35e5df99b015.png)

alisonelizabeth commented on x-pack/plugins/painless_lab/server/routes/api/execute.ts:
```diff
@@ -12,7 +12,16 @@ import { isEsError } from '../../lib';
+    schema.object({
+      params: schema.maybe(schema.any()),
+      document: schema.recordOf(schema.string(), schema.any()),
+      index: schema.string(),
```
I made this required, since it looks like ES requires it, but I also opened https://github.com/elastic/kibana/issues/60020 as a follow-up.

alisonelizabeth commented:
@cjcenizal great catch! I have updated the schema, would you mind taking another look and testing again? I've confirmed adding `params` is working as expected now. I've tried to test changing the context, but my browser keeps freezing up (using [this example](https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-execute-api.html#_filter_context)). I tested on the feature branch as well and experience the same behavior, so I don't _think_ it's related to this PR. Have you run into this before?

I also noticed a couple other issues:
- The `index` field appears to be required by the API when changing the context, but there is no indication in the UI (opened https://github.com/elastic/kibana/issues/60020)
- Documentation links are hard-coded throughout the code (opened https://github.com/elastic/kibana/issues/60024)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32815/)
* Commit: e87cf919b2f42ac5c9addc103fc295047e24ce3d

### History
* :green_heart: [Build #32569](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32569/) succeeded 54b5d15aacd2c10f520a846acaf9eef0fd187fb1
* :green_heart: [Build #32415](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32415/) succeeded c76542fe057b8a2b8701ca403a3f3815251b936c
* :green_heart: [Build #32269](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/) succeeded c0c0b8e0a5d56fe2db59964686d2cdf907d3d008
* :broken_heart: [Build #32200](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/) failed d5cf2b358b8e06d02b32cd766f41079bf2f2c188
* :broken_heart: [Build #32164](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/) failed b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32164/","number":"32164","commit":"b681a1861b7c7e24e7a7ed469bf5ae90b2c6b2d1"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32200/","number":"32200","commit":"d5cf2b358b8e06d02b32cd766f41079bf2f2c188"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32269/","number":"32269","commit":"c0c0b8e0a5d56fe2db59964686d2cdf907d3d008"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32415/","number":"32415","commit":"c76542fe057b8a2b8701ca403a3f3815251b936c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32569/","number":"32569","commit":"54b5d15aacd2c10f520a846acaf9eef0fd187fb1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32815/","number":"32815","commit":"e87cf919b2f42ac5c9addc103fc295047e24ce3d"}]}
PIPELINE-->
